src/go/rpk/pkg/cli/transform/deploy.go

```diff
@@ -0,0 +1,242 @@
+	var fc deployFlagConfig
+
+	cmd := &cobra.Command{
+		Use:   "deploy [WASM]",
+		Short: "Deploy a transform",
+		Long: `Deploy a transform.
+
+When run in the same directory as a transform.yaml it will read the configuration file,
+then look for a .wasm file with the same name as your project. If the input and output
+topics are specified in the configuration file, those will be used, otherwise the topics
+can be specified on the command line using the --input-topic and --output-topic flags.
+
+Wasm files can also be deployed directly without a transform.yaml file by specifying it
+like so:
+
+rpk transform deploy transform.wasm --name myTransform
+
+Environment variables can be specified for the transform using the --env-var flag, these
+are separated by an equals for example: --env-var=KEY=VALUE
+
+The --env-var can be repeated to specify multiple variables.
+		`,
+		Args: cobra.MaximumNArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			p, err := p.LoadVirtualProfile(fs)
+			out.MaybeDie(err, "unable to load config: %v", err)
+
+			api, err := adminapi.NewClient(fs, p)
+			out.MaybeDie(err, "unable to initialize admin api client: %v", err)
+
+			cfg := fc.ToProjectConfig()
+			fileConfig, err := project.LoadCfg(fs)
+			// We allow users to deploy if they aren't in a directory with transform.yaml
+			// in that case all config needs to be specified on the command line.
+			if err != nil {
+				cfg = mergeProjectConfigs(fileConfig, cfg)
+			}
+			err = validateProjectConfig(cfg, err)
+			out.MaybeDieErr(err)
+
+			deployable := fmt.Sprintf("%s.wasm", cfg.Name)
+			if len(args) == 1 {
+				deployable = args[0]
+			}
+			wasm, err := loadWasm(fs, deployable)
+			out.MaybeDieErr(err)
+
+			t := adminapi.TransformMetadata{
+				InputTopic:   cfg.InputTopic,
+				OutputTopics: []string{cfg.OutputTopic},
+				Name:         cfg.Name,
+				Status:       nil,
+				Environment:  mapToEnvVars(cfg.Env),
+			}
+			err = api.DeployWasmTransform(cmd.Context(), t, wasm)
+			if he := (*adminapi.HTTPResponseError)(nil); errors.As(err, &he) {
+				if he.Response.StatusCode == 400 {
+					body, bodyErr := he.DecodeGenericErrorBody()
+					if bodyErr == nil {
+						out.Die("unable to deploy transform %s: %s", cfg.Name, body.Message)
+					}
+				}
+			}
+			out.MaybeDie(err, "unable to deploy transform %s: %v", cfg.Name, err)
+
+			fmt.Printf("transform %q deployed.\n", cfg.Name)
+		},
+	}
+	cmd.Flags().StringVarP(&fc.inputTopic, "input-topic", "i", "", "The input topic to apply the transform to")
+	cmd.Flags().StringVarP(&fc.outputTopic, "output-topic", "o", "", "The output topic to write the transform results to")
+	cmd.Flags().StringVar(&fc.functionName, "name", "", "The name of the transform")
+	cmd.Flags().Var(&fc.env, "env-var", "Specify an environment variable in the form of KEY=VALUE")
+	return cmd
+}
+
+type environment struct {
+	vars map[string]string
+}
+
+func (e *environment) Set(s string) error {
```

Comment:
Whoop! Glad it's helpful. TBH I mostly did the custom flag type to prevent having to keep a `[]string` around and convert it to a `map[string]string` later. Moving that to flag parse time simplifies the other logic.